mica/stats/mag_stats/catalogs.py

```diff
@@ -0,0 +1,228 @@
+import os
+import sys
+import numpy as np
+
+from astropy.table import Table
+from astropy import table
+from Ska.DBI import DBI
+
+from cxotime import CxoTime
+import tables
+from Chandra.Time import date2secs
+
+TIMELINES = None
+STARCAT_CMDS = None
+DWELLS_NP = None
+DWELLS_MAP = None
+STARS_OBS = None
+STARS_OBS_INDICES = None
+STARS_OBS_MAP = None
+STARS_OBS_NP = None
+
+
+def _load_startcat_commands(tstop=None):
+    """
+    Get all star catalog commands.  This is a definitive list of commands
+    *actually run* on the spacecraft (to the best of our knowledge).
+
+    Note this one case in the cmds archive of duplicate commands.  Not clear
+    if this is real or an artifact of bad timelines.
+
+    <Table length=7>
+     idx            date            type     tlmsid   scs   step  timeline_id   vcdu
+    uint16        bytes21         bytes12   bytes10  uint8 uint16    uint32    int32
+    ------ --------------------- ---------- -------- ----- ------ ----------- --------
+        4 2006:295:18:59:08.748 MP_STARCAT AOSTRCAT   128     25   277930627   181961
+        4 2006:295:18:59:08.748 MP_STARCAT AOSTRCAT   130    706   277867085   181961
+        6 2006:295:18:59:12.999 COMMAND_SW AOMANUVR   130    757   277867085   181977
+        6 2006:295:18:59:12.999 COMMAND_SW AOMANUVR   128     76   277930627   181977
+
+    :return:
+    """
+    from kadi import events
+    from kadi.commands import get_cmds
+
+    global STARCAT_CMDS
+    global DWELLS_NP
+    global DWELLS_MAP
+
+    STARCAT_CMDS = Table(get_cmds('2003:001', type='MP_STARCAT'))['date', 'timeline_id']
+    STARCAT_CMDS.rename_column('date', 'mp_starcat_time')
+    # print(len(starcat_cmds))
+    STARCAT_CMDS = table.unique(STARCAT_CMDS, keys='mp_starcat_time')
+    # print(len(starcat_cmds))
+
+    # Get all maneuvers and use this to create a table of NPM dwells
+    # with a star catalog.
+    #
+    # Note that obsid is not unique for maneuvers since
+    # after a safing action the obsid in telemetry can stay fixed while maneuvers
+    # continue happening.
+    manvrs = events.manvrs.filter('2003:001')
+    rows = [(manvr.obsid, manvr.start, manvr.kalman_start, manvr.npnt_stop) for manvr in manvrs]
+
+    # Form dwells table. At this point it includes maneuvers that might not have
+    # a star catalog with NPM dwell.  That gets fixed in the next bit.
+    dwells = Table(rows=rows, names=('obsid', 'manvr_start', 'start', 'stop'))
```

Comment:
Since everybody else is defining a dwell from npnt start to npnt stop, might want to leave the "kalman_start" label hanging around?